src/mscorlib/shared/System/Buffers/ValueListBuilder.cs

```diff
@@ -0,0 +1,66 @@
+﻿using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers
+{
+    internal ref struct ValueListBuilder<T>
+    {
+        private Span<T> _span;
+        private T[] _arrayFromPool;
+        private int _pos;
+
+        public ValueListBuilder(Span<T> initialSpan)
+        {
+            _span = initialSpan;
+            _arrayFromPool = null;
+            _pos = 0;
+        }
+
+        public int Length => _pos;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Append(T item)
+        {
+            int pos = _pos;
+            if (pos < _span.Length)
+            {
+                _span[pos] = item;
+                _pos = pos + 1;
+            }
+            else
+            {
+                Grow();
+                Append(item);
```

Comment:
It is better to implement like this without recursive call:

```
if (pos >= _span.Length)
    Grow();

_span[pos] = item;
_pos = pos + 1;
```

Reply:
definitely, thank you

Reply:
> It is better to implement like this without recursive call

Why? I could be misremembering, but I thought I'd looked at the asm for both and the fast path was better this way. Maybe not.

Reply:
The code generators tend to convert recursive calls like these into loops.

If you have verified by looking at the disassembly that this gives the best code, it is fine.

Reply:
I could be wrong and would need to look again.